### PR TITLE
Updates `traversable` to test the correct laws

### PR DESCRIPTION
**Pull request by solomon-b**
Resolves issue #24

I wasn't able to implement the Naturality test correctly, but I left it in as a comment in case someone else can get it working. I think it follows from Identity and Composition so we are okay without it.

### CODE REVIEW

File: src/Test/QuickCheck/Classes.hs
Comment (sjakobi) [inline src/Test/QuickCheck/Classes.hs:742]:
```suggestion
   -- FIXME: Doesn't compile.
   --naturalityP :: (forall x. (f x -> g x)) -> (a -> f b) -> Property
   --naturalityP t f = t . traverse @t f =-= traverse (t . f)
```

(Feel free to improve on the explanation)

File: src/Test/QuickCheck/Classes.hs
Comment (sjakobi) [inline src/Test/QuickCheck/Classes.hs:739]: Please break this line.

File: src/Test/QuickCheck/Classes.hs
Comment (sjakobi) [inline src/Test/QuickCheck/Classes.hs:728]: What's the reason for removing these laws? Are they implied by identity and composition?

I noticed that both `quickcheck-classes` and `hedgehog-classes` test these laws too, so I'm inclined to keep them.

**Review by sjakobi — COMMENTED** [review]
Thanks for your PR! :)

> I think it follows from Identity and Composition so we are okay without it.

Do you know a proof?

I'm fine with scrapping support for GHC 7.10 BTW, so you can use `TypeApplications`.

**Comment by sjakobi** [comment]
BTW, it would be nice to have tests for the `sequenceA` laws, too. Feel free to add them, but there's no obligation. :)

File: src/Test/QuickCheck/Classes.hs
Comment (solomon-b) [reply src/Test/QuickCheck/Classes.hs:728]: Sorry, that was a mistake! I'll bring them back.

**Comment by solomon-b** [comment]
> > I think it follows from Identity and Composition so we are okay without it.
> 
> Do you know a proof?

hmm, i actually haven't been able to find a proof, but this link claims it is guaranteed to hold:
https://en.wikibooks.org/wiki/Haskell/Traversable#The_Traversable_laws

**Review by sjakobi — COMMENTED** [review]
Thanks!

One more nit…

File: src/Test/QuickCheck/Classes.hs
Comment (solomon-b) [reply src/Test/QuickCheck/Classes.hs]: done!

**Review by sjakobi — APPROVED** [review]
Thanks! :)

Since this is a breaking change, I'll merge this after cutting the v0.5.7 release (#64).

**Comment by solomon-b** [comment]
Great!

**Comment by sjakobi** [comment]
Regarding naturality, note this change: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7492

**Comment by solomon-b** [comment]
Well there you go. Should we hold off on merging this until a working implementation is produced?

**Comment by sjakobi** [comment]
> Well there you go. Should we hold off on merging this until a working implementation is produced?

When do you expect it to land?

IMHO the existing changes are good additions which shouldn't be held back for very long.

**Comment by solomon-b** [comment]
> > Well there you go. Should we hold off on merging this until a working implementation is produced?
> 
> When do you expect it to land?
> 
> IMHO the existing changes are good additions which shouldn't be held back for very long.

To be honest, I'm a little stumped on this how to fix it. Is there anyone with more quickcheck experience who we can ask? The issue is the rank2 type.

**Comment by sjakobi** [comment]
Published in v0.6.0: https://hackage.haskell.org/package/checkers-0.6.0

**Pull request merged**
